junebug/amqp.py

```diff
@@ -0,0 +1,115 @@
+from twisted.internet.defer import inlineCallbacks, returnValue, Deferred
+from twisted.internet.protocol import ReconnectingClientFactory
+from twisted.python import log
+from txamqp.client import TwistedDelegate
+from txamqp.content import Content
+from txamqp.protocol import AMQClient
+from vumi.utils import vumi_resource_path
+from vumi.service import get_spec
+
+
+class AmqpFactory(ReconnectingClientFactory, object):
+    def __init__(self, specfile, amqp_config):
+        '''Factory that creates JunebugAMQClients.
+        specfile - string of specfile name
+        amqp_config - connection details for amqp server
+        '''
+        self.amqp_config = amqp_config
+        self.spec = get_spec(vumi_resource_path(specfile))
+        self.delegate = TwistedDelegate()
+        self.amqp_client_d = Deferred()
+
+    def buildProtocol(self, addr):
+        amqp_client = JunebugAMQClient(
+            self.delegate, self.amqp_config['vhost'],
+            self.spec, self.amqp_config.get('heartbeat', 0))
+        amqp_client.factory = self
+        self.resetDelay()
+        return amqp_client
+
+    def clientConnectionFailed(self, connector, reason):
+        log.err("AmqpFactory connection failed (%s)" % (
+            reason.getErrorMessage(),))
+        super(AmqpFactory, self).clientConnectionFailed(connector, reason)
+
+    def clientConnectionLost(self, connector, reason):
+        log.err("AmqpFactory client connection lost (%s)" % (
+            reason.getErrorMessage(),))
+        self.amqp_client_d = Deferred()
+        super(AmqpFactory, self).clientConnectionLost(connector, reason)
+
+    @inlineCallbacks
+    def get_client(self):
+        '''Returns a deferred that fires with a connected client'''
+        if not self.amqp_client_d.called:
+            yield self.amqp_client_d
+        returnValue(self.amqp_client)
```

Comment:
From the slack discussion (making a comment here so I don't forget), the plan is to get the client class via a class level variable, and then override that class level variable with a dummy client class and use that as a way to test `AmqpFactory`.